Overhaul "Defining a schema" documentation (#3662)
ChilliCream/hotchocolate: website/src/docs/hotchocolate/defining-a-schema/extending-types.md

import { ExampleTabs } from "../../../components/mdx/example-tabs"

GraphQL types tend to become pretty large, especially root types like the query type. Type extensions allow us to extend existing types.

A type can have one or more type extensions, which will form a combined schema type at runtime.

# Usage

Given is the following entity that we want to extend with functionality.

```csharp
public class Book
{
    public int Id { get; set; }

    public string Title { get; set; }

    public int AuthorId { get; set; }
}

// This is only relevant for the Code-first approach
public class BookType : ObjectType<Book>
{
}
```

## Adding fields

We can easily add new fields to an existing type.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[ExtendObjectType(typeof(Book))]
public class BookExtensions
{
    public IEnumerable<string> GetGenres([Parent] Book book)
    {
        // Omitted code for brevity
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Code-first adding new fields to an existing type

</ExampleTabs.Code>
<ExampleTabs.Schema>

Simply add a new field to the existing type.

</ExampleTabs.Schema>
</ExampleTabs>

One of the most common use-cases for this would be adding new resolvers to one of our root types.

```csharp
[ExtendObjectType(typeof(Query))]
public class QueryBookResolvers
{
    public IEnumerable<Book> GetBooks()
    {
        // Omitted code for brevity
    }
}
```

## Removing fields

We can also ignore fields of the type we are extending.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp

[ExtendObjectType(typeof(Book),
    IgnoreProperties = new[] { nameof(Book.AuthorId) })]
public class BookExtensions
{
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Code-first removing fields from an existing type

</ExampleTabs.Code>
<ExampleTabs.Schema>

Simply remove the field from the existing type.

</ExampleTabs.Schema>
</ExampleTabs>

## Replacing fields

We might have an `Id` field, which we want to replace with a field that resolves the actual type the `Id` is pointing to.

In this example we replace the `authorId` field with an `author` field.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[ExtendObjectType(typeof(Book))]
public class BookExtensions
{
    [BindMember(nameof(Book.AuthorId))]
    public Author GetAuthor([Parent] Book book)
    {
        // Omitted code for brevity
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Code-first replacing existing field on existing type

Reviewer: Doing this by targeting a base type is super powerful and something that'd be great to document with some real world examples. 

Personally we use this heavily in return payload types and just normal types to pull in things by just their Id. Meaning you get CQRS for free on your payloads and stitching becomes a breeze. We extend a baste type like a `IHasBlogPost` or `ICanHaveBlogPost` with the resolver to look it up by id, and using the dataloader, and then annotate all kinds of things with that to do local stitching (e.g. when not remote stitching) and the Q in CQRS, because our mutation can just set the ids on the return type and then the resolver will do a fresh fetch of it for us. 

If you need some examples, here's the original issue, though Michael's comment has all the finalised syntax really.
https://github.com/ChilliCream/hotchocolate/issues/2991

import { ExampleTabs } from "../../../components/mdx/example-tabs"

GraphQL types tend to become pretty large, especially root types like the query type. Type extensions allow us to extend existing types.

A type can have one or more type extensions, which will form a combined schema type at runtime.

# Usage

Given is the following entity that we want to extend with functionality.

```csharp
public class Book
{
    public int Id { get; set; }

    public string Title { get; set; }

    public int AuthorId { get; set; }
}

// This is only relevant for the Code-first approach
public class BookType : ObjectType<Book>
{
}
```

## Adding fields

We can easily add new fields to an existing type.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[ExtendObjectType(typeof(Book))]
public class BookExtensions
{
    public IEnumerable<string> GetGenres([Parent] Book book)
    {
        // Omitted code for brevity
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Add Code-first

</ExampleTabs.Code>
<ExampleTabs.Schema>

> TODO

Author: Add Schema-first

```csharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddDocumentFromString(@"
                type Book {
                  id: Int!
                  title: String!
                  authorId: Int!
                }

                extend type Book {
                  genres: [String!]!
                }
            ");
    }
}
```

</ExampleTabs.Schema>
</ExampleTabs>

One of the most common use-cases for this would be adding new resolvers to one of our root types.

> <!-- todo: maybe with example tabs -->

Author: Decide this and get rid of it

```csharp
[ExtendObjectType(typeof(Query))]
public class QueryBookResolvers
{
    public IEnumerable<Book> GetBooks()
    {
        // Omitted code for brevity
    }
}
```

## Removing fields

We can also ignore fields of the type we are extending.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp

[ExtendObjectType(typeof(Book),
    IgnoreProperties = new[] { nameof(Book.AuthorId) })]
public class BookExtensions
{
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Add Code-first

</ExampleTabs.Code>
<ExampleTabs.Schema>

Simply remove the field from the original type.

</ExampleTabs.Schema>
</ExampleTabs>

## Replacing fields

We might have an `Id` field, which we want to replace with a field that resolves the actual type the `Id` is pointing to.

In this example we replace the `authorId` field with an `author` field.

<ExampleTabs>
<ExampleTabs.Annotation>

```csharp
[ExtendObjectType(typeof(Book))]
public class BookExtensions
{
    [BindMember(nameof(Book.AuthorId))]
    public Author GetAuthor([Parent] Book book)
    {
        // Omitted code for brevity
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookExtensions>();
    }
}
```

</ExampleTabs.Annotation>
<ExampleTabs.Code>

> TODO

Author: Add Code-first

}
```

In Hot Chocolate type extensions work a little different. Instead of generating the `extend` syntax in the schema, the original type definition and its type extensions are merged at runtime into a combined type.

Type extensions make most sense, if

- we want to split up types into separate classes.

- we can not modify the original type.

> Note: If we just want to organize the fields of one of our types in different files, we can use [partial classes](https://docs.microsoft.com/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods) in the Annotation-based approach.

# Object Types

> Given is the following entity that we want to extend with functionality.

Reviewer: ```suggestion
Consider we have the following entity that we want to extend with functionality.
```

```csharp
public class Book
{
    public int Id { get; set; }

    public string Title { get; set; }

    public int AuthorId { get; set; }
}
```

## Adding fields

> We can easily add new fields to an existing type.

Reviewer: ```suggestion
We can easily add new fields to our existing `Booking` type.
```

Reviewer: I was trying to have it flow from the top, but maybe this is overkill. Happy as you had it before too

            {
                var parent = context.Parent<Book>();

                // Omitted code for brevity
            });
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services
            .AddGraphQLServer()
            .AddTypeExtension<BookTypeExtensions>();
    }
}
```

</ExampleTabs.Code>
<ExampleTabs.Schema>

Simply replace the field on the existing type.

</ExampleTabs.Schema>
</ExampleTabs>

## Extending base types

We can also extend multiple types at once, but still dedicate specific resolvers to specific types.

```csharp
// this extends every type that inherits from object (essentially every type)
[ExtendObjectType(typeof(object))]
public class ObjectExtensions
{
    // this field is added to every object type
    public string NewField()
    {
        // Omitted code for brevity
    }

    // this field is only added to the Book type
    public Author GetAuthor([Parent] Book book)
    {
        // Omitted code for brevity
    }

    // this field is only added to the Author type
    public IEnumerable<Book> GetBooks([Parent] Author author)
    {
        // Omitted code for brevity
    }
}
```

> We can also modify all object types that are connected by an interface.

Reviewer: ```suggestion
We can also modify all object types that are connected by a base type, like an interface.

> Note: Below `IPost` is annotated with `[InterfaceType]` to include it in the GraphQL schema, but that isn't necessary for the type extension to work. You can use any base type, like `object`, an `abstract` base class, as an extension point without necessarily exposing the base type in your GraphQL schema.
```

Perhaps the note is unnecessary if the attribute is dropped from the sample, but just wanted to make it clearer that it's not req.
